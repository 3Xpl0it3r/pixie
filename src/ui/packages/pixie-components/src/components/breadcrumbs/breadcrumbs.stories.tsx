/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as React from 'react';

import { StatusCell } from 'components/status/status';
import { Breadcrumbs } from './breadcrumbs';

export default {
  title: 'Breadcrumbs',
  component: Breadcrumbs,
  subcomponents: { StatusCell },
  decorators: [
    (Story) => (
      <div style={{ backgroundColor: '#212324' }}>
        <Story />
      </div>
    ),
  ],
};

export const Basic = () => {
  const breadcrumbs = [
    {
      title: 'cluster',
      value: 'gke-prod',
      selectable: true,
      allowTyping: false,
      // eslint-disable-next-line
      getListItems: async () => [
        // Description is optional, so only providing it on some to demonstrate what happens when it isn't provided.
        {
          value: 'cluster1',
          icon: <StatusCell statusGroup='healthy' />,
          description: 'Cluster 1 description',
        },
        {
          value: 'cluster2',
          icon: <StatusCell statusGroup='unhealthy' />,
          description: 'Cluster 2 description',
        },
        { value: 'cluster3', icon: <StatusCell statusGroup='pending' /> },
      ],
      onSelect: (input) => {
        // eslint-disable-next-line
        console.log(`Selected cluster: ${input}`);
      },
    },
    {
      title: 'pod',
      value: 'pod-123',
      selectable: true,
      allowTyping: true,
      getListItems: async (input) => {
        if (input === '') {
          return [
            { value: 'pod1' },
            {
              value: 'pod2',
              description: 'Pod 2 has a description, 1 does not',
            },
          ];
        }
        return [
          { value: 'some pod' },
          { value: 'another pod', description: 'Interrupting cow says what?' },
          { value: 'pod' },
        ];
      },
      onSelect: (input) => {
        // eslint-disable-next-line
        console.log(`Selected pod: ${input}`);
      },
    },
    {
      title: 'script',
      value: 'px/pod',
      selectable: false,
    },
  ];

  return <Breadcrumbs breadcrumbs={breadcrumbs} />;
};
