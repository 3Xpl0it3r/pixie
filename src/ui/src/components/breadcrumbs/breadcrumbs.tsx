import * as React from 'react';
import {
  createStyles, Theme, WithStyles, withStyles, ThemeProvider,
} from '@material-ui/core/styles';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import { Card, Popover } from '@material-ui/core';
import createTypography from '@material-ui/core/styles/createTypography';
import createSpacing from '@material-ui/core/styles/createSpacing';
import { CompletionItem } from 'components/autocomplete/completions';
import Autocomplete from 'components/autocomplete';
import { AutocompleteContext } from 'components/autocomplete/autocomplete';

const styles = ({ spacing, typography, palette }: Theme) => createStyles({
  breadcrumbs: {
    height: '100%',
    overflowX: 'scroll',
    scrollbarWidth: 'none', // Firefox
    '&::-webkit-scrollbar': {
      display: 'none',
    },
    display: 'flex',
  },
  breadcrumb: {
    display: 'inline-flex',
    paddingRight: spacing(0.5),
    alignItems: 'center',
  },
  title: {
    fontWeight: typography.fontWeightMedium,
    paddingRight: spacing(0.5),
    paddingLeft: spacing(0.5),
    fontFamily: '"Roboto Mono", Monospace',
  },
  card: {
    width: '608px',
  },
  autocomplete: {
    maxHeight: '60vh',
  },
  value: {
    color: palette.primary.main,
    whiteSpace: 'nowrap',
    fontFamily: '"Roboto Mono", Monospace',
  },
  body: {
    ...typography.body2,
    display: 'inline-block',
    color: palette.text.primary,
    height: spacing(3),
  },
  content: {
    display: 'flex',
    alignItems: 'center',
    paddingLeft: spacing(1),
    height: spacing(3),
  },
  dropdownArrow: {
    cursor: 'pointer',
    height: spacing(3),
  },
  spacer: {
    width: spacing(2),
  },
  input: {
    padding: `0 ${spacing(1)}px ${spacing(1)}px ${spacing(1)}px`,
  },
  completionsContainer: {
    // 80% as wide as the Command Input, and using 80% of the fontSize and 80% of the spacing
    maxWidth: '608px',
    maxHeight: '60vh',
  },
  // CSS for the front/back of the breadcrumb element.
  triangle: {
    position: 'absolute',
    height: spacing(3),
    width: spacing(3),
    overflow: 'hidden',
  },
  angle: {
    float: 'right', // Take the triangle out of the DOM, so it can overlap with other breadcrumbs.
  },
  tail: {
    width: spacing(2.4),
    height: spacing(3),
  },
  topArrow: {
    position: 'absolute',
    width: spacing(3),
    height: spacing(3),
    top: -1 * spacing(1.5),
    right: -1 * spacing(1.6),
    background: palette.foreground.grey3,
    transform: 'rotate(45deg)',
    borderRadius: spacing(0.4),
  },
  bottomArrow: {
    position: 'absolute',
    width: spacing(3),
    height: spacing(3),
    bottom: `-${spacing(1.5)}px`,
    right: `-${spacing(1.6)}px`,
    background: palette.foreground.grey3,
    transform: 'rotate(45deg)',
    borderRadius: spacing(0.4),
  },
  arrow: {
    position: 'absolute',
    width: spacing(3),
    height: spacing(3),
    left: `-${spacing(1.5)}px`,
    background: palette.foreground.grey3,
    transform: 'rotate(45deg)',
    borderRadius: spacing(0.4),
  },
  inputItem: {
    '&:hover': {
      backgroundColor: 'initial',
    },
    '&.Mui-focusVisible': {
      backgroundColor: 'initial',
    },
  },
  separator: {
    display: 'flex',
    alignItems: 'center',
    color: palette.foreground.one,
    fontWeight: 1000,
    width: spacing(1),
  },
  icon: {
    minWidth: spacing(4),
  },
});

interface DialogDropdownProps extends WithStyles<typeof styles> {
  onSelect: (input: string) => void;
  getListItems: (input: string) => Promise<BreadcrumbListItem[]>;
  onClose: () => void;
  anchorEl: HTMLElement;
}

const DialogDropdown = ({
  classes, onSelect, onClose, getListItems, anchorEl,
}: DialogDropdownProps) => {
  const { allowTyping, requireCompletion, preSelect } = React.useContext(AutocompleteContext);
  const [completionItems, setCompletionItems] = React.useState<CompletionItem[]>([]);
  const inputRef = React.useRef<HTMLInputElement>(null);

  if (anchorEl && inputRef.current) {
    setTimeout(() => {
      inputRef.current?.focus();
    }, 100);
  }

  const onCompletionSelected = React.useCallback((itemValue: string) => {
    if (requireCompletion) {
      // Completion required, refresh the valid options and verify that the user picked one of them.
      // As getListItems is async, we refresh it so that the check is run atomically and can't be out of date.
      if (typeof getListItems !== 'function') throw new Error(`List items not gettable when selecting "${itemValue}"!`);
      getListItems(itemValue).then((items) => {
        if (items.length < 1) throw new Error('Failed to match the input to a valid choice!');
        onSelect(itemValue);
        onClose();
      });
    } else {
      // Completion not required, so we don't care if the user picked an item that actually exists. Skip validation.
      onSelect(itemValue);
      onClose();
    }
  }, [requireCompletion, getListItems, onClose, onSelect]);

  const getCompletions = React.useCallback((input: string) => {
    if (typeof getListItems === 'function') {
      return getListItems(input).then((items) => {
        const mapped: CompletionItem[] = items.map((item) => ({
          type: 'item' as 'item',
          id: item.value,
          description: item.description,
          icon: item.icon,
          title: item.value,
        }));
        setCompletionItems(mapped);
        return mapped;
      });
    }

    setCompletionItems([]);
    return Promise.resolve([]);
  }, [getListItems]);

  // Used to shrink the <Autocomplete/>'s fonts and negative space, to not be as huge / central as the Command Input.
  const compactThemeBuilder = React.useMemo<(theme: Theme) => Theme>(() => (theme: Theme) => ({
    ...theme,
    typography: createTypography(theme.palette, { fontSize: theme.typography.fontSize * 0.8 }),
    spacing: createSpacing((factor) => (factor * 0.8 * theme.spacing(1))),
  }), []);

  return (
    <Popover
      classes={{ paper: classes.completionsContainer }}
      anchorEl={anchorEl}
      keepMounted
      open={Boolean(anchorEl)}
      onClose={onClose}
      getContentAnchorEl={null}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'left',
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'left',
      }}
      elevation={0}
    >
      <ThemeProvider theme={compactThemeBuilder}>
        <Card className={classes.card}>
          <AutocompleteContext.Provider value={{
            allowTyping,
            requireCompletion,
            inputRef,
            preSelect: !!(preSelect && anchorEl),
          }}
          >
            <Autocomplete
              className={classes.autocomplete}
              placeholder='Filter...'
              onSelection={onCompletionSelected}
              getCompletions={getCompletions}
            />
          </AutocompleteContext.Provider>
        </Card>
      </ThemeProvider>
    </Popover>
  );
};

interface BreadcrumbProps extends WithStyles<typeof styles> {
  title: string;
  value: string;
  selectable: boolean;
  getListItems?: (input: string) => Promise<Array<BreadcrumbListItem>>;
  onSelect?: (input: string) => void;
  omitKey?: boolean;
}

const Breadcrumb = ({
  classes, title, value, selectable, getListItems, onSelect, omitKey,
}: BreadcrumbProps) => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  const handleClick = React.useCallback((event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget.parentElement);
  }, [setAnchorEl]);

  const onClose = React.useCallback(() => {
    setAnchorEl(null);
  }, [setAnchorEl]);

  return (
    <div className={classes.breadcrumb}>
      <div className={classes.body}>
        <div className={classes.content}>
          {!omitKey && <span className={classes.title}>{`${title}: `}</span>}
          <span className={classes.value}>{value}</span>
          {selectable && <div className={classes.dropdownArrow} onClick={handleClick}><ArrowDropDownIcon /></div>}
          {!selectable && <div className={classes.spacer} />}
          <DialogDropdown
            classes={classes}
            onSelect={onSelect}
            onClose={onClose}
            getListItems={getListItems}
            anchorEl={anchorEl}
          />
        </div>
      </div>
    </div>
  );
};

interface BreadcrumbListItem {
  value: string;
  description?: string;
  icon?: React.ReactNode;
}

export interface BreadcrumbOptions {
  title: string;
  value: string;
  selectable: boolean;
  allowTyping?: boolean;
  getListItems?: (input: string) => Promise<Array<BreadcrumbListItem>>;
  onSelect?: (input: string) => void;
  requireCompletion?: boolean;
}

interface BreadcrumbsProps extends WithStyles<typeof styles> {
  breadcrumbs: BreadcrumbOptions[];
}

// TODO(nserrino/michelle): Support links (non-menu) as a type of breadcrumb,
// replace breadcrumbs in cluster details page with that new type of breadcrumb.
const Breadcrumbs = ({
  classes, breadcrumbs,
}: BreadcrumbsProps) => {
  // In case a breadcrumb doesn't override, give it the nearest context's values.
  const { allowTyping, requireCompletion } = React.useContext(AutocompleteContext);
  return (
    <div className={classes.breadcrumbs}>
      {
        breadcrumbs.map((breadcrumb, i) => (
          // Key is needed to prevent a console error when a key is missing in a list element.
          <AutocompleteContext.Provider
            key={`i-${breadcrumb.title}`}
            value={{
              allowTyping: breadcrumb.allowTyping ?? allowTyping,
              requireCompletion: breadcrumb.requireCompletion ?? requireCompletion,
              preSelect: true,
            }}
          >
            <Breadcrumb
              key={i}
              classes={classes}
              {...breadcrumb}
            />
          </AutocompleteContext.Provider>
        ))
      }
    </div>
  );
};

export default withStyles(styles)(Breadcrumbs);
