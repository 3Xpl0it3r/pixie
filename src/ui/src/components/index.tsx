/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export { AuthBox } from './auth/auth-box';
export { GoogleButton } from './auth/google-button';
export { UsernamePasswordButton } from './auth/username-password-button';
export {
  Form, FormField, FormStructure, PixienautForm,
} from './form/form';
export { AuthMessageBoxProps, AuthMessageBox } from './auth/message';
export {
  PixienautBoxProps,
  PixienautBox,
} from './auth/pixienaut-box';
export { SignupMarcom } from './auth/signup-marcom';

export { Autocomplete } from './autocomplete/autocomplete';
export { AutocompleteContextProps, AutocompleteContext } from 'components/autocomplete/autocomplete-context';
export {
  CompletionHeader,
  CompletionId,
  CompletionTitle,
  CompletionItem,
  CompletionItems,
  Completion,
  Completions,
} from './autocomplete/completions';
export { FormFieldInput } from './autocomplete/form';
export { FormInput, Input } from './autocomplete/input';
export {
  AutocompleteField,
  CommandAutocompleteInput,
} from 'components/autocomplete/command-autocomplete-input';
export {
  CommandAutocomplete,
  TabSuggestion,
} from 'components/autocomplete/command-autocomplete';
export {
  useAutocomplete,
  GetCompletionsFunc,
} from './autocomplete/use-autocomplete';
export {
  ItemsMap,
  findNextItem,
  TabStop,
  getDisplayStringFromTabStops,
  TabStopParser,
} from './autocomplete/utils';

export {
  BreadcrumbOptions,
  Breadcrumbs,
} from './breadcrumbs/breadcrumbs';

export { CodeEditor } from './code-editor/code-editor';

export { CodeRenderer } from './code-renderer/code-renderer';

export {
  CellAlignment,
  ColumnProps,
  ExpandedRows,
  SortState,
  DataTable,
} from './data-table/data-table';
export { ColWidthOverrides } from './data-table/table-resizer';

export { FixedSizeDrawer } from './drawer/drawer';
export { ResizableDrawer } from './drawer/resizable-drawer';

export { Footer } from './footer/footer';

export { LazyPanel } from './lazy-panel/lazy-panel';

export { ModalTrigger } from './modal/modal';

export { Avatar, ProfileMenuWrapper } from './profile/profile';

export { Select } from './select/select';

export {
  SelectedPercentile,
  QuantilesBoxWhisker,
} from './quantiles-box-whisker/quantiles-box-whisker';

export { SnackbarContext, SnackbarProvider, useSnackbar } from './snackbar/snackbar';

export { Spinner } from './spinner/spinner';

export { SplitContainer, SplitPane } from './split-pane/split-pane';

export { StatusGroup, StatusCell } from './status/status';

export { VersionInfo } from './version-info/version-info';

export { buildClass } from 'utils/build-class';

export { ClusterIcon } from './icons/cluster';
export { CodeIcon } from './icons/code';
export { CopyIcon } from './icons/copy';
export { DocsIcon } from './icons/docs';
export { EditIcon } from './icons/edit';
export { GoogleIcon } from './icons/google';
export { LogoutIcon } from './icons/logout';
export { MagicIcon } from './icons/magic';
export { NamespaceIcon } from './icons/namespace';
export { PlayIcon } from './icons/play';
export { StopIcon } from './icons/stop';
export { PodIcon } from './icons/pod';
export { ServiceIcon } from './icons/service';
export { SettingsIcon } from './icons/settings';

export { PixieCommandIcon } from './icons/pixie-command';
export { PixieCommandHint } from './icons/pixie-command-hint';
export { PixieLogo } from './icons/pixie-logo';

export { scrollbarStyles, DARK_THEME } from './mui-theme';
