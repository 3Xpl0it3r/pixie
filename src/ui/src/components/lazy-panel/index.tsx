import LazyPanel from './lazy-panel';

export default LazyPanel;
