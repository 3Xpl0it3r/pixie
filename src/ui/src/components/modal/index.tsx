import * as root from './modal';

export default root.ModalTrigger;
