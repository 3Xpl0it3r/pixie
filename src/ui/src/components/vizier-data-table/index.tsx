import {VizierDataTable} from './vizier-data-table';

export default VizierDataTable;
