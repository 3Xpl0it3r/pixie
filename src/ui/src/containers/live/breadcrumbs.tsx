import * as React from 'react';

import {
  createStyles, Theme, withStyles,
} from '@material-ui/core/styles';
import { useQuery, useApolloClient } from '@apollo/client';

import {
  Breadcrumbs, BreadcrumbOptions,
  PixieCommandIcon, StatusCell,
} from '@pixie/components';
import { ClusterContext } from 'common/cluster-context';
import { CLUSTER_STATUS_DISCONNECTED } from 'common/vizier-grpc-client-context';
import {
  getArgTypesForVis, getArgVariableMap,
} from 'utils/args-utils';
import { SCRATCH_SCRIPT, ScriptsContext } from 'containers/App/scripts-context';
import { ScriptContext } from 'context/script-context';
import { entityPageForScriptId, optionallyGetNamespace } from 'containers/live-widgets/utils/live-view-params';
import { EntityType, pxTypetoEntityType, entityStatusGroup } from 'containers/command-input/autocomplete-utils';
import { clusterStatusGroup } from 'containers/admin/utils';
import { containsMutation, CLUSTER_QUERIES, AUTOCOMPLETE_QUERIES } from '@pixie/api';
import ExecuteScriptButton from './execute-button';
import { Variable } from './vis';

const styles = (({ shape, palette, spacing }: Theme) => createStyles({
  root: {
    display: 'flex',
    alignItems: 'center',
    marginTop: spacing(1),
    marginRight: spacing(3.5),
    marginLeft: spacing(3),
    marginBottom: spacing(1),
    background: palette.background.three,
    // This adds a scroll to the breadcrumbs on overflow,
    // but it's hard for the user to know it exists. Perhaps we can
    // consider adding a scroll effect or something to make it easier to
    // discover.
    borderRadius: shape.borderRadius,
    boxShadow: '4px 4px 4px rgba(0, 0, 0, 0.25)',
    border: palette.border.unFocused,
  },
  spacer: {
    flex: 1,
  },
  verticalLine: {
    borderLeft: `2px solid ${palette.foreground.grey1}`,
    height: spacing(2.7),
    padding: 0,
  },
  pixieIcon: {
    color: palette.primary.main,
    marginRight: spacing(1),
    marginLeft: spacing(0.5),
  },
  iconContainer: {
    padding: 0,
  },
  breadcrumbs: {
    width: '100%',
    overflow: 'hidden',
    display: 'flex',
  },
}));

const LiveViewBreadcrumbs = ({ classes }) => {
  const { loading, data } = useQuery(CLUSTER_QUERIES.LIST_CLUSTERS);
  const { selectedCluster, setCluster, selectedClusterUID } = React.useContext(ClusterContext);
  const { scripts } = React.useContext(ScriptsContext);

  const {
    vis, pxl, args, id, liveViewPage, setArgs, execute, setScript, parseVisOrShowError, argsForVisOrShowError,
  } = React.useContext(ScriptContext);

  const client = useApolloClient();
  const getCompletions = React.useCallback((newInput: string, kind: EntityType) => (client.query({
    query: AUTOCOMPLETE_QUERIES.FIELD,
    fetchPolicy: 'network-only',
    variables: {
      input: newInput,
      kind,
      clusterUID: selectedClusterUID,
    },
  })
  ), [client, selectedClusterUID]);

  const scriptIds = React.useMemo(() => [...scripts.keys()], [scripts]);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const clusterIds = React.useMemo(() => data?.clusters.map((c) => c.id), [data?.clusters]);

  type MemoCrumbs = { entityBreadcrumbs: BreadcrumbOptions[]; argBreadcrumbs: BreadcrumbOptions[] };
  const { entityBreadcrumbs, argBreadcrumbs }: MemoCrumbs = React.useMemo(() => {
    // eslint-disable-next-line no-shadow
    const entityBreadcrumbs: BreadcrumbOptions[] = [];
    // eslint-disable-next-line no-shadow
    const argBreadcrumbs: BreadcrumbOptions[] = [];
    if (loading) return { entityBreadcrumbs, argBreadcrumbs };
    // Cluster always goes first in breadcrumbs.
    const clusterName = data.clusters.find((c) => c.id === selectedCluster)?.prettyClusterName || 'unknown cluster';
    const clusterNameToID: Record<string, string> = {};
    data.clusters.forEach((c) => {
      clusterNameToID[c.prettyClusterName] = c.id;
    });
    entityBreadcrumbs.push({
      title: 'cluster',
      value: clusterName,
      selectable: true,
      // eslint-disable-next-line
      getListItems: async (input) => (data.clusters.filter((c) => c.status !== CLUSTER_STATUS_DISCONNECTED
              && c.prettyClusterName.includes(input))
        .map((c) => ({ value: c.prettyClusterName, icon: <StatusCell statusGroup={clusterStatusGroup(c.status)} /> }))
      ),
      onSelect: (input) => {
        setCluster(clusterNameToID[input]);
      },
      requireCompletion: true,
    });

    // Add args to breadcrumbs.
    const argTypes = getArgTypesForVis(vis);

    const argVariableMap = getArgVariableMap(vis);

    // TODO(michelle): We may want to separate non-entity args from the entity args and put them in separate
    // breadcrumbs. For now, they will all go in the same breadcrumbs object.
    Object.entries(args).filter(([argName]) => argName !== 'script').forEach(([argName, argVal]) => {
      // Only add suggestions if validValues are specified. Otherwise, the dropdown is populated with autocomplete
      // entities or has no elements and the user must manually type in values.
      const variable: Variable = argVariableMap[argName];

      const argProps = {
        title: argName,
        value: argVal.toString(),
        selectable: true,
        allowTyping: true,
        onSelect: (newVal) => {
          const newArgs = { ...args, [argName]: newVal };
          setArgs(newArgs);
          execute({
            pxl, vis, args: newArgs, id, liveViewPage,
          });
        },
        getListItems: null,
        requireCompletion: false,
        placeholder: variable?.description,
      };

      if (variable && typeof variable.defaultValue === 'undefined') {
        argProps.title += '*';
      }

      if (variable?.validValues?.length) {
        argProps.getListItems = async (input) => (variable.validValues
          .filter((suggestion) => input === '' || suggestion.indexOf(input) >= 0)
          .map((suggestion) => ({
            value: suggestion,
            description: '',
          })));

        argProps.requireCompletion = true;
      }

      const entityType = pxTypetoEntityType(argTypes[argName]);
      if (entityType !== 'AEK_UNKNOWN') {
        argProps.getListItems = async (input) => (getCompletions(input, entityType)
          .then((results) => (results.data.autocompleteField.map((suggestion) => ({
            value: suggestion.name,
            description: suggestion.description,
            icon: <StatusCell statusGroup={entityStatusGroup(suggestion.state)} />,
          })))));
      }

      // TODO(michelle): Ideally we should just be able to use the entityType to determine whether the
      // arg appears in the argBreadcrumbs. However, some entities still don't have a corresponding entity type
      // (such as nodes), since they are not yet supported in autocomplete. Until that is fixed, this is hard-coded
      // for now.
      if (argName === 'start_time' || argName === 'start') {
        argBreadcrumbs.push(argProps);
      } else {
        entityBreadcrumbs.push(argProps);
      }
    });

    // Add script at end of breadcrumbs.
    entityBreadcrumbs.push({
      title: 'script',
      value: id,
      selectable: true,
      allowTyping: true,
      getListItems: async (input) => {
        // Turns "  px/be_spoke-  " into "px/bespoke"
        const normalize = (s: string) => s.trim().toLowerCase().replace(/[^a-z0-9/]+/gi, '');
        const normalizedInput = normalize(input);
        const normalizedScratchId = normalize(SCRATCH_SCRIPT.id);

        const ids = !input ? [...scriptIds] : scriptIds.filter((s) => {
          const ns = normalize(s);
          return ns === normalizedScratchId || ns.indexOf(normalizedInput) >= 0;
        });

        // The scratch script should always appear at the top of the list for visibility. It doesn't get auto-selected
        // unless it's the only thing in the list.
        const scratchIndex = scriptIds.indexOf(SCRATCH_SCRIPT.id);
        if (scratchIndex !== -1) {
          scriptIds.splice(scratchIndex, 1);
          scriptIds.unshift(SCRATCH_SCRIPT.id);
        }

        return ids.map((scriptId) => ({
          value: scriptId,
          description: scripts.get(scriptId).description,
          autoSelectPriority: scriptId === SCRATCH_SCRIPT.id ? -1 : 0,
        }));
      },
      onSelect: (newVal) => {
        const script = scripts.get(newVal);
        const selectedVis = parseVisOrShowError(script.vis);
        const parsedArgs = argsForVisOrShowError(selectedVis, {
          // Grab the namespace if it exists on a service or pod argument.
          namespace: optionallyGetNamespace(args),
          ...args,
        });
        if (!selectedVis && !parsedArgs) {
          return;
        }

        const execArgs = {
          liveViewPage: entityPageForScriptId(newVal),
          pxl: script.code,
          id: newVal,
          args: parsedArgs,
          vis: selectedVis,
        };
        setScript(execArgs.vis, execArgs.pxl, execArgs.args, execArgs.id, execArgs.liveViewPage);
        if (newVal === SCRATCH_SCRIPT.id) {
          // Skip executing the script, which starts empty. Executing empty scripts emits scary (harmless) errors.
          return;
        }
        if (!containsMutation(execArgs.pxl)) {
          execute(execArgs);
        }
      },
    });

    return { entityBreadcrumbs, argBreadcrumbs };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [loading, id, clusterIds?.join(','), scriptIds?.join(','), args, selectedCluster, getCompletions]);

  if (loading) {
    return (<div>Loading...</div>);
  }

  return (
    <div className={classes.root}>
      <PixieCommandIcon fontSize='large' className={classes.pixieIcon} />
      <div className={classes.verticalLine} />
      <div className={classes.breadcrumbs}>
        <Breadcrumbs
          breadcrumbs={entityBreadcrumbs}
        />
        <div className={classes.spacer} />
        <div>
          <Breadcrumbs
            breadcrumbs={argBreadcrumbs}
          />
        </div>
      </div>
      <ExecuteScriptButton />
    </div>
  );
};

export default withStyles(styles)(LiveViewBreadcrumbs);
