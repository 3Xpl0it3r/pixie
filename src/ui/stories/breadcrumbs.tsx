import * as React from 'react';

import { Breadcrumbs, StatusCell } from 'pixie-components';

export default {
  title: 'Breadcrumbs',
  component: Breadcrumbs,
  subcomponents: { StatusCell },
  decorators: [(Story) => <div style={{ backgroundColor: '#212324' }}><Story /></div>],
};

export const Basic = () => {
  const breadcrumbs = [
    {
      title: 'cluster',
      value: 'gke-prod',
      selectable: true,
      allowTyping: false,
      // eslint-disable-next-line
      getListItems: async () => (
        [
          // Description is optional, so only providing it on some to demonstrate what happens when it isn't provided.
          { value: 'cluster1', icon: <StatusCell statusGroup='healthy' />, description: 'Cluster 1 description' },
          { value: 'cluster2', icon: <StatusCell statusGroup='unhealthy' />, description: 'Cluster 2 description' },
          { value: 'cluster3', icon: <StatusCell statusGroup='pending' /> },
        ]
      ),
      onSelect: (input) => {
        // eslint-disable-next-line
        console.log(`Selected cluster: ${input}`);
      },
    },
    {
      title: 'pod',
      value: 'pod-123',
      selectable: true,
      allowTyping: true,
      getListItems: async (input) => {
        if (input === '') {
          return [{ value: 'pod1' }, { value: 'pod2', description: 'Pod 2 has a description, 1 does not' }];
        }
        return [
          { value: 'some pod' },
          { value: 'another pod', description: 'Interrupting cow says what?' },
          { value: 'pod' }];
      },
      onSelect: (input) => {
        // eslint-disable-next-line
        console.log(`Selected pod: ${input}`);
      },
    },
    {
      title: 'script',
      value: 'px/pod',
      selectable: false,
    },
  ];

  return <Breadcrumbs breadcrumbs={breadcrumbs} />;
};
