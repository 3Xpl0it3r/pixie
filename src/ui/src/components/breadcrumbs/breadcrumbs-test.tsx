/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from 'enzyme';
import { ThemeProvider } from '@material-ui/core/styles';
import { DARK_THEME } from 'components';
import { Breadcrumbs } from './breadcrumbs';

describe('<Breadcrumbs/>', () => {
  it('renders correctly', () => {
    const breadcrumbs = [
      {
        title: 'cluster',
        value: 'gke-prod',
        selectable: true,
        allowTyping: false,
        getListItems: async (input) => {
          if (input) {
            return [
              { value: 'cluster1' },
              { value: 'cluster2' },
              { value: 'cluster3' },
            ];
          }
          return [
            { value: 'cluster1' },
            { value: 'cluster2' },
            { value: 'cluster3' },
          ];
        },
      },
      {
        title: 'pod',
        value: 'pod-123',
        selectable: true,
        allowTyping: true,
        getListItems: async (input) => {
          if (input === '') {
            return [{ value: 'pod1' }, { value: 'pod2' }];
          }
          return [
            { value: 'some pod' },
            { value: 'another pod' },
            { value: 'pod' },
          ];
        },
      },
      {
        title: 'script',
        value: 'px/pod',
        selectable: false,
      },
    ];

    const wrapper = render(
      <ThemeProvider theme={DARK_THEME}>
        <Breadcrumbs breadcrumbs={breadcrumbs} />
      </ThemeProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
