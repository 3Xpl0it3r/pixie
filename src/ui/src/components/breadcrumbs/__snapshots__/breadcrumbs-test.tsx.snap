// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs/> renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1459qpu-MuiPaper-root"
  >
    <div
      class="BreadcrumbsImpl-breadcrumbs-1"
    >
      <div
        class="BreadcrumbsImpl-breadcrumb-2"
      >
        <div
          class="BreadcrumbsImpl-body-7"
        >
          <div
            class="BreadcrumbsImpl-content-8"
          >
            <span
              class="BreadcrumbsImpl-title-3"
            >
              cluster: 
            </span>
            <span
              class="BreadcrumbsImpl-value-6"
            >
              gke-prod
            </span>
            <div
              class="BreadcrumbsImpl-dropdownArrow-9"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m7 10 5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="BreadcrumbsImpl-breadcrumb-2"
      >
        <div
          class="BreadcrumbsImpl-body-7"
        >
          <div
            class="BreadcrumbsImpl-content-8"
          >
            <span
              class="BreadcrumbsImpl-title-3"
            >
              pod: 
            </span>
            <span
              class="BreadcrumbsImpl-value-6"
            >
              pod-123
            </span>
            <div
              class="BreadcrumbsImpl-dropdownArrow-9"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m7 10 5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="BreadcrumbsImpl-breadcrumb-2"
      >
        <div
          class="BreadcrumbsImpl-body-7"
        >
          <div
            class="BreadcrumbsImpl-content-8"
          >
            <span
              class="BreadcrumbsImpl-title-3"
            >
              script: 
            </span>
            <span
              class="BreadcrumbsImpl-value-6"
            >
              px/pod
            </span>
            <div
              class="BreadcrumbsImpl-spacer-10"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
